OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/fieldassignability/L2FieldAssignabilityAnalysis.scala

```diff
@@ -149,30 +147,59 @@ class L2FieldAssignabilityAnalysis private[analyses] (val project: SomeProject)
         val writes = state.fieldWriteAccessDependee.get.ub.accesses
         val writesInMethod = writes.filter { w => contextProvider.contextFromId(w._1).method eq definedMethod }.toSeq
 
-        if (writesInMethod.distinctBy(_._2).size > 1)
-            return true; // Field is written in multiple locations, thus must be assignable
+        if (writesInMethod.distinctBy(_._2).size > 1) {
+            // There can be multiple assignments of final fields in the constructor
+            // in different branches
+            if (!definedMethod.definedMethod.isConstructor)
+                return true
+        }; // Otherwise: field is written in multiple locations, thus must be assignable
 
-        // If we have no information about the receiver, we soundly return
-        if (receiverVar.isEmpty)
+        // If we have no information about the receiver, we soundly return true
+        // However, a static field has no receiver
+        if (receiverVar.isEmpty && !state.field.isStatic)
             return true;
 
-        val assignedValueObject = receiverVar.get
-        if (assignedValueObject.definedBy.exists(_ < 0))
+        val assignedValueObject =
+            if (index > 0 && stmts(index).isPutStatic) {
+                stmts(index).asPutStatic.value.asVar
+            } else if (
+                index > 0 && stmts(index).isPutField && stmts(index).asPutField.value.asVar.value.isArrayValue.isYes
```

Comment:
Please document this case, it is unclear why it is like this. In particular, why do we take the value from all PutStatic instructions, but only PutField instructions with array values? And why can't we rely on the receiverVar in all cases? Shouldn't it be same the asPut{Static,Field}.value.asVar anyway?